src/dependencies/useStoreDependency.ts

```diff
@@ -0,0 +1,82 @@
+import * as React from 'react';
+import {
+  Dependency,
+  calculate,
+  calculateForDispatch,
+  makeDependencyIndex,
+  DependencyIndexEntry,
+} from './DependencyMap';
+import { Dispatcher } from 'flux';
+import { get as getDispatcherInstance } from '../dispatcher/DispatcherInstance';
+import { enforceDispatcher } from '../dispatcher/DispatcherInterface';
+import { handleDispatch } from './Dispatch';
+
+type SingleDependency = {
+  [key: string]: Dependency;
+};
+
+function useStoreDependency<Props>(
+  dependency: Dependency,
+  props: Props,
+  dispatcher: Dispatcher<any> = getDispatcherInstance()
+) {
+  enforceDispatcher(dispatcher);
+
+  const dispatchToken = React.useRef(null);
+  const { current: dependencyMap } = React.useRef<SingleDependency>({
+    dependency,
+  });
+  const { current: dependencyIndex } = React.useRef(
+    makeDependencyIndex(dependencyMap)
+  );
+  const [dependencyValue, setDependencyValue] = React.useState(
+    calculate(dependency, props)
+  );
+
+  const propsArray = Object.keys(props || {})
+    .sort()
+    .map(key => props[key]);
+
+  React.useEffect(() => {
+    dispatchToken.current = dispatcher.register(
+      handleDispatch.bind(
+        null,
+        dispatcher,
+        dependencyIndex,
+        (entry: DependencyIndexEntry) => {
+          const {
+            dependency: newValue,
+          }: SingleDependency = calculateForDispatch(
+            dependencyMap,
+            entry,
+            props
+          );
+          if (newValue !== dependencyValue) {
+            setDependencyValue(newValue);
+          }
+        }
+      )
+    );
+    return () => {
+      dispatcher.unregister(dispatchToken.current);
+      dispatchToken.current = null;
+    };
+  }, [
+    dispatcher,
+    dependencyIndex,
+    dispatchToken,
+    dependencyValue,
+    dependencyMap,
+  ]);
+
+  React.useMemo(() => {
```

Comment:
it's surprising to see a `useMemo` hook not returning a value. Is the idea here to conditionally trigger a side-effect during the render?

Reply:
If you run into a warning like this: ` React Hook React.useMemo was passed a dependency list that is not an array literal. This means we can't statically verify whether you've passed the correct dependencies` I'm thinking this might be better as something like this:
```
  const prevProps = usePrevious(props);
  useEffect(() => {
     if (!shallowEqual(props, prevProps)) {
      const newValue = calculate(dependency, props);
      if (newValue !== dependencyValue) {
        setDependencyValue(newValue);
      }
     }

  }, [props]);
```

Reply:
after revisiting this, i think we might actually just want this to be at the top level of the hook, not inside of a hook at all. on every render that means some of the parent components's props/state changed, which means we almost definitely want to do the recalculation at that point. it doesn't appear to be causing extra re-renders, so i think just having the following at the bottom of the hook is sufficient for what i want here:

```js
const newValue = calculate(dependency, props);
if (newValue !== dependencyValue) {
  setDependencyValue(newValue);
}
```